server/src/main/java/io/crate/operation/collect/files/CSVLineParser.java

```diff
@@ -33,16 +33,32 @@
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
 
 public class CSVLineParser {
 
-    private final ArrayList<String> keyList = new ArrayList<>();
     private final ByteArrayOutputStream out = new ByteArrayOutputStream();
+    private final Map<Integer, String> keyPositions = new HashMap<>();
```

Comment:
```suggestion
    private final Map<Integer, String> columnNameByFieldIndex = new HashMap<>();
```

Could also use a IntObjectMap to avoid the boxing, or even an array with gaps to reduce memory usage even further.

Reply:
1. If we use an array to keep positions wouldn't we have to preallocate an array of maximum column size even for object references?
Something like columnNameByFieldArray[MAXIMUM_ALLOWED_COLUMNS]?

2. Btw i saw that default table columns is 1000 but can be changed through mapping.total_fields.limit. Is there any absolute limit for column size? 

If not use something like 
columnNameByFieldArray[Integer.MAX_VALUE - 8]?

Reply:
I wasn't aware of IntObjectHashMap class and its efficiency for primitive types. Thanks!
I also saw an android library of SparseArray that is an array with gaps but I didn't find something similar in standard library.

What would you suggest If I wanted to use an array with gaps?

Reply:
I made a different implementation with an array but had to have two iterations inside parseHeader method. As it is only triggered once per csv file I think that the overhead isn't significant.